File: control-plane/src/test/java/it/unimib/datai/nanofaas/controlplane/execution/ExecutionRecordStateTransitionTest.java
package it.unimib.datai.nanofaas.controlplane.execution;

import it.unimib.datai.nanofaas.common.model.ErrorInfo;
import it.unimib.datai.nanofaas.controlplane.scheduler.InvocationTask;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

class ExecutionRecordStateTransitionTest {

    @Test
    void validTransition_queued_to_running() {
        ExecutionRecord record = createRecord("exec-1");
        assertThat(record.state()).isEqualTo(ExecutionState.QUEUED);

        record.markRunning();
        assertThat(record.state()).isEqualTo(ExecutionState.RUNNING);
    }

    @Test
    void validTransition_running_to_success() {
        ExecutionRecord record = createRecord("exec-1");
        record.markRunning();

        record.markSuccess("output");
        assertThat(record.state()).isEqualTo(ExecutionState.SUCCESS);
        assertThat(record.output()).isEqualTo("output");
    }

    @Test
    void validTransition_running_to_error() {
        ExecutionRecord record = createRecord("exec-1");
        record.markRunning();

        ErrorInfo error = new ErrorInfo("TEST_ERROR", "something failed");
        record.markError(error);
        assertThat(record.state()).isEqualTo(ExecutionState.ERROR);
        assertThat(record.lastError()).isEqualTo(error);
    }

    @Test
    void validTransition_running_to_timeout() {
        ExecutionRecord record = createRecord("exec-1");
        record.markRunning();

        record.markTimeout();
        assertThat(record.state()).isEqualTo(ExecutionState.TIMEOUT);
    }

    @Test
    void validTransition_running_to_queued_viaResetForRetry() {
        ExecutionRecord record = createRecord("exec-1");
        record.markRunning();

        InvocationTask retryTask = createTask("exec-1");
        record.resetForRetry(retryTask);
        assertThat(record.state()).isEqualTo(ExecutionState.QUEUED);
        assertThat(record.startedAt()).isNull();
        assertThat(record.finishedAt()).isNull();
    }

    @Test
    void invalidTransition_queued_to_success_logsWarning() {
        // Should not throw but logs warning - transition still happens
        ExecutionRecord record = createRecord("exec-1");
        assertThat(record.state()).isEqualTo(ExecutionState.QUEUED);

        record.markSuccess("output");
        // State still changes (warn-only)
        assertThat(record.state()).isEqualTo(ExecutionState.SUCCESS);
    }

    @Test
    void invalidTransition_success_to_running_logsWarning() {
        ExecutionRecord record = createRecord("exec-1");
        record.markRunning();
        record.markSuccess("output");

        record.markRunning();
        // State still changes (warn-only), SUCCESS -> RUNNING is invalid
        assertThat(record.state()).isEqualTo(ExecutionState.RUNNING);
    }

    @Test
    void invalidTransition_error_to_success_logsWarning() {
        ExecutionRecord record = createRecord("exec-1");
        record.markRunning();
        record.markError(new ErrorInfo("ERR", "failed"));

        record.markSuccess("output");
        // State still changes (warn-only)
        assertThat(record.state()).isEqualTo(ExecutionState.SUCCESS);
    }

    @Test
    void snapshot_returnsConsistentView() {
        ExecutionRecord record = createRecord("exec-1");
        record.markRunning();

        ExecutionRecord.Snapshot snapshot = record.snapshot();
        assertThat(snapshot.executionId()).isEqualTo("exec-1");
        assertThat(snapshot.state()).isEqualTo(ExecutionState.RUNNING);
        assertThat(snapshot.startedAt()).isNotNull();
        assertThat(snapshot.finishedAt()).isNull();
    }

    @Test
    void markColdStart_setsFieldsInSnapshot() {
        ExecutionRecord record = createRecord("exec-1");
        record.markRunning();
        record.markColdStart(350);

        ExecutionRecord.Snapshot snapshot = record.snapshot();
        assertThat(snapshot.coldStart()).isTrue();
        assertThat(snapshot.initDurationMs()).isEqualTo(350L);
    }

    @Test
    void markDispatchedAt_setsFieldInSnapshot() {
        ExecutionRecord record = createRecord("exec-1");
        record.markRunning();
        record.markDispatchedAt();

        ExecutionRecord.Snapshot snapshot = record.snapshot();
        assertThat(snapshot.dispatchedAt()).isNotNull();
    }

    @Test
    void resetForRetry_clearsColdStartFields() {
        ExecutionRecord record = createRecord("exec-1");
        record.markRunning();
        record.markColdStart(200);
        record.markDispatchedAt();

        record.resetForRetry(createTask("exec-1"));

        ExecutionRecord.Snapshot snapshot = record.snapshot();
        assertThat(snapshot.coldStart()).isFalse();
        assertThat(snapshot.initDurationMs()).isNull();
        assertThat(snapshot.dispatchedAt()).isNull();
    }

    private ExecutionRecord createRecord(String executionId) {
        return new ExecutionRecord(executionId, createTask(executionId));
    }

    private InvocationTask createTask(String executionId) {
        return new InvocationTask(executionId, "testFunc", null, null, null, null, null, 1);
    }
}

File: experiments/control-plane-staging/versions/control-plane-rust-m3-20260222-200159/snapshot/control-plane-rust/tests/execution_record_state_transition_test.rs
#![allow(non_snake_case)]

use control_plane_rust::execution::{ErrorInfo, ExecutionRecord, ExecutionState, InvocationTask};

#[test]
fn validTransition_queued_to_running() {
    let mut record = create_record("exec-1");
    assert_eq!(record.state(), ExecutionState::Queued);

    record.mark_running_at(10);
    assert_eq!(record.state(), ExecutionState::Running);
}

#[test]
fn validTransition_running_to_success() {
    let mut record = create_record("exec-1");
    record.mark_running_at(10);

    record.mark_success_at(serde_json::json!("output"), 20);
    assert_eq!(record.state(), ExecutionState::Success);
    assert_eq!(record.output(), Some(serde_json::json!("output")));
}

#[test]
fn validTransition_running_to_error() {
    let mut record = create_record("exec-1");
    record.mark_running_at(10);

    let error = ErrorInfo::new("TEST_ERROR", "something failed");
    record.mark_error_at(error.clone(), 20);
    assert_eq!(record.state(), ExecutionState::Error);
    assert_eq!(record.last_error(), Some(error));
}

#[test]
fn validTransition_running_to_timeout() {
    let mut record = create_record("exec-1");
    record.mark_running_at(10);

    record.mark_timeout_at(20);
    assert_eq!(record.state(), ExecutionState::Timeout);
}

#[test]
fn validTransition_running_to_queued_viaResetForRetry() {
    let mut record = create_record("exec-1");
    record.mark_running_at(10);

    let retry_task = create_task("exec-1");
    record.reset_for_retry(retry_task);
    assert_eq!(record.state(), ExecutionState::Queued);
    assert_eq!(record.started_at_millis(), None);
    assert_eq!(record.finished_at_millis(), None);
}

#[test]
fn invalidTransition_queued_to_success_logsWarning() {
    let mut record = create_record("exec-1");
    assert_eq!(record.state(), ExecutionState::Queued);

    record.mark_success_at(serde_json::json!("output"), 10);
    assert_eq!(record.state(), ExecutionState::Success);
}

#[test]
fn invalidTransition_success_to_running_logsWarning() {
    let mut record = create_record("exec-1");
    record.mark_running_at(10);
    record.mark_success_at(serde_json::json!("output"), 20);

    record.mark_running_at(30);
    assert_eq!(record.state(), ExecutionState::Running);
}

#[test]
fn invalidTransition_error_to_success_logsWarning() {
    let mut record = create_record("exec-1");
    record.mark_running_at(10);
    record.mark_error_at(ErrorInfo::new("ERR", "failed"), 20);

    record.mark_success_at(serde_json::json!("output"), 30);
    assert_eq!(record.state(), ExecutionState::Success);
}

#[test]
fn snapshot_returnsConsistentView() {
    let mut record = create_record("exec-1");
    record.mark_running_at(10);

    let snapshot = record.snapshot();
    assert_eq!(snapshot.execution_id, "exec-1");
    assert_eq!(snapshot.state, ExecutionState::Running);
    assert_eq!(snapshot.started_at_millis, Some(10));
    assert_eq!(snapshot.finished_at_millis, None);
}

#[test]
fn markColdStart_setsFieldsInSnapshot() {
    let mut record = create_record("exec-1");
    record.mark_running_at(10);
    record.mark_cold_start(350);

    let snapshot = record.snapshot();
    assert!(snapshot.cold_start);
    assert_eq!(snapshot.init_duration_ms, Some(350));
}

#[test]
fn markDispatchedAt_setsFieldInSnapshot() {
    let mut record = create_record("exec-1");
    record.mark_running_at(10);
    record.mark_dispatched_at(11);

    let snapshot = record.snapshot();
    assert_eq!(snapshot.dispatched_at_millis, Some(11));
}

#[test]
fn resetForRetry_clearsColdStartFields() {
    let mut record = create_record("exec-1");
    record.mark_running_at(10);
    record.mark_cold_start(200);
    record.mark_dispatched_at(11);

    record.reset_for_retry(create_task("exec-1"));

    let snapshot = record.snapshot();
    assert!(!snapshot.cold_start);
    assert_eq!(snapshot.init_duration_ms, None);
    assert_eq!(snapshot.dispatched_at_millis, None);
}

fn create_record(execution_id: &str) -> ExecutionRecord {
    ExecutionRecord::new_with_task(execution_id, create_task(execution_id))
}

fn create_task(execution_id: &str) -> InvocationTask {
    InvocationTask::new(execution_id, "testFunc", 1)
}
